src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala

```diff
@@ -0,0 +1,144 @@
+        val main_contract_ast = mainContract match {
+            case obsContract: ObsidianContractImpl => translateContract(obsContract)
+            case _ => throw new RuntimeException("Java contract not supported in yul translation")
+        }
+
+        // TODO ignore imports, data for now
+        // translate other contracts (if any) and add them to the subObjects
+        var new_subObjects: Seq[YulObject] = main_contract_ast.subObjects
+        for (c <- program.contracts) {
+            c match {
+                case obsContract: ObsidianContractImpl =>
+                    if (!c.modifiers.contains(IsMain())) { // if not main contract
+                        new_subObjects = main_contract_ast.subObjects :+ translateContract(obsContract)
+                    }
+                case _: JavaFFIContractImpl =>
+                    throw new RuntimeException("Java contract not supported in yul translation")
+            }
+        }
+        YulObject(main_contract_ast.name, main_contract_ast.code,
+            new_subObjects, main_contract_ast.data)
+    }
+
+    def translateContract(contract: ObsidianContractImpl): YulObject = {
+        // create runtime obj
+        var subObjects: Seq[YulObject] = Seq()
+        val runtime_name = contract.name + "_deployed"
+        val runtime_obj = YulObject(runtime_name, Code(Block(Seq())), Seq(), Seq())
+        subObjects = runtime_obj +: subObjects
+
+        var statement_seq: Seq[YulStatement] = Seq()
+        // TODO initialization
+
+        val create = YulObject(contract.name, Code(Block(statement_seq)), subObjects, Seq())
+        return create
+    }
+
+
+    def translateDeclaration(declaration: Declaration): Seq[YulStatement] = {
+        declaration match {
+            case f: Field => translateField(f)
+            case t: Transaction =>
+                assert(false, "TODO")
+                Seq()
+            case s: State =>
+                assert(false, "TODO")
+                Seq()
+            case c: ObsidianContractImpl =>
+                assert(false, "TODO")
+                Seq()
+            case c: JavaFFIContractImpl =>
+                assert(false, "TODO")
+                Seq()
+            case c: Constructor =>
+                assert(false, "TODO")
+                Seq()
+            case t: TypeDecl =>
+                assert(false, "TODO")
+                Seq()
+            // This should never be hit.
+            case _ =>
+                assert(false, "Translating unexpected declaration: " + declaration)
+                Seq()
+        }
+    }
+
+    def translateField(field: Field): Seq[YulStatement] = {
+        // Reserve a slot in the storage by assigning a index in the symbol table
+        // since field declaration has not yet be assigned, there is no need to do sstore
+        temp_symbol_table += field.name -> temp_table_idx
+        temp_table_idx += 1
+        Seq()
+    }
+
+    def yulString(obj: YulObject, finalOutputPath: Path)= {
```

Comment:
This should not have side effects. It should instead return a string, which the caller can write to a path. The principle here is that methods whose names do not indicate side effects should not have any (in this case, writing a file).